File: ts/project/.sdk/tm/rb/feature/clienttrack_feature.rb
# ProjectName SDK clienttrack feature
#
# Client tracking. Establishes a stable per-client session id at
# construction and stamps identifying headers on every request: a
# "User-Agent" ("<clientName>/<clientVersion>"), an "X-Client-Id"
# (session), and a fresh per-request "X-Request-Id". This lets a server
# correlate all traffic from one SDK instance and each individual call.
# Header names, client name/version and the id generator ("idgen") are
# configurable; caller-provided User-Agent / X-Client-Id values are never
# clobbered.

require_relative 'base_feature'

class ProjectNameClienttrackFeature < ProjectNameBaseFeature
  def initialize
    super
    @version = "0.0.1"
    @name = "clienttrack"
    # Inactive until init (feature_init only fires init when active).
    @active = false
    @client = nil
    @options = {}
    @session = ""
    @requests = 0
  end

  def init(ctx, options)
    @client = ctx.client
    @options = options.is_a?(Hash) ? options : {}
    @active = @options["active"] == true
    @requests = 0
  end

  def PostConstruct(ctx)
    return unless @active
    @session = @options["sessionId"] || _genid("session")
    @client.instance_variable_set(:@_clienttrack, {
      "session" => @session,
      "requests" => 0,
      "clientName" => _client_name,
    })
  end

  def PreRequest(ctx)
    return unless @active
    spec = ctx.spec
    return if spec.nil?
    spec.headers = {} if spec.headers.nil?
    if @session == ""
      @session = @options["sessionId"] || _genid("session")
    end

    h = @options["headers"] || {}
    @requests += 1
    request_id = _genid("request")

    _set(spec.headers, h["agent"] || "User-Agent", _client_name)
    _set(spec.headers, h["client"] || "X-Client-Id", @session)
    spec.headers[h["request"] || "X-Request-Id"] = request_id

    track = @client.instance_variable_get(:@_clienttrack)
    if track.nil?
      track = { "session" => @session, "requests" => 0, "clientName" => _client_name }
      @client.instance_variable_set(:@_clienttrack, track)
    end
    track["requests"] = @requests
    track["lastRequestId"] = request_id
  end

  # Do not clobber a caller-provided value (e.g. a custom User-Agent).
  def _set(headers, name, value)
    lower = name.downcase
    headers.each_key do |k|
      return if k.to_s.downcase == lower
    end
    headers[name] = value
  end

  def _client_name
    name = @options["clientName"] || "ProjectName-SDK"
    version = @options["clientVersion"] || "0.0.1"
    "#{name}/#{version}"
  end

  def _genid(kind)
    idgen = @options["idgen"]
    return idgen.call(kind) if idgen.is_a?(Proc)
    h = -> { rand(0x10000000).to_s(16) }
    "#{kind[0]}-#{h.call}#{h.call}#{h.call}"[0, 20]
  end
end

File: ts/project/.sdk/tm/py/feature/clienttrack_feature.py
# ProjectName SDK clienttrack feature

from __future__ import annotations
import random

from feature.base_feature import ProjectNameBaseFeature


# Client tracking. Establishes a stable per-client session id at
# construction and stamps identifying headers on every request: a
# `User-Agent`, an `X-Client-Id` (session), and a fresh per-request
# `X-Request-Id`. This lets a server correlate all traffic from one SDK
# instance and each individual call. Header names, client name/version
# (`clientName`/`clientVersion`) and the id generator (`idgen`) are
# configurable; the session id and request counter are exposed on
# `client._clienttrack`. Caller-provided User-Agent / X-Client-Id headers
# are never clobbered.
class ProjectNameClienttrackFeature(ProjectNameBaseFeature):
    def __init__(self):
        super().__init__()
        self.version = "0.0.1"
        self.name = "clienttrack"
        self.active = True
        self.client = None
        self.options = {}
        self.session = ""
        self.requests = 0

    def init(self, ctx, options):
        self.client = ctx.client
        self.options = options if isinstance(options, dict) else {}

        if self.options.get("active") is True:
            self.active = True
        else:
            self.active = False

        self.requests = 0

    def PostConstruct(self, ctx):
        if not self.active:
            return
        self.session = self.options.get("sessionId") or self._genid("session")
        self.client._clienttrack = {
            "session": self.session,
            "requests": 0,
            "clientName": self._name(),
        }

    def PreRequest(self, ctx):
        if not self.active:
            return

        spec = ctx.spec
        if spec is None:
            return
        if spec.headers is None:
            spec.headers = {}
        if self.session == "":
            self.session = self.options.get("sessionId") or self._genid("session")

        headers = self.options.get("headers") or {}
        self.requests += 1
        request_id = self._genid("request")

        self._set(spec.headers, headers.get("agent") or "User-Agent", self._name())
        self._set(spec.headers, headers.get("client") or "X-Client-Id", self.session)
        spec.headers[headers.get("request") or "X-Request-Id"] = request_id

        client = self.client
        track = getattr(client, "_clienttrack", None)
        if track is None:
            track = {"session": self.session, "requests": 0, "clientName": self._name()}
            client._clienttrack = track
        track["requests"] = self.requests
        track["lastRequestId"] = request_id

    # Do not clobber a caller-provided value (e.g. a custom User-Agent).
    def _set(self, headers, name, value):
        lower = name.lower()
        for key in headers:
            if str(key).lower() == lower:
                return
        headers[name] = value

    def _name(self):
        name = self.options.get("clientName") or "ProjectName-SDK"
        version = self.options.get("clientVersion") or "0.0.1"
        return name + "/" + version

    def _genid(self, kind):
        idgen = self.options.get("idgen")
        if callable(idgen):
            return idgen(kind)
        return (kind[0] + "-" + "%04x%04x%04x%04x" % (
            random.randint(0, 0xFFFF), random.randint(0, 0xFFFF),
            random.randint(0, 0xFFFF), random.randint(0, 0xFFFF)))[:20]
